File: app/src/main/java/org/peercast/core/PeerCastService.java
package org.peercast.core;

/**
 * (c) 2014, T Yoshizawa
 * Dual licensed under the MIT or GPL licenses.
 */

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.res.AssetManager;
import android.os.Bundle;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.IBinder;
import android.os.Looper;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;
import android.util.Log;

import org.apache.commons.io.FileUtils;
import org.apache.commons.io.IOUtils;
import org.peercast.pecaport.PecaPortService;

import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

import static org.peercast.core.NotificationHelper.ACTION_BUMP;
import static org.peercast.core.NotificationHelper.ACTION_DISCONNECT;


public class PeerCastService extends Service {
    private static final String TAG = "PeerCastService";

    public static final int MSG_GET_APPLICATION_PROPERTIES = 0x00;
    public static final int MSG_GET_CHANNELS = 0x01;
    public static final int MSG_GET_STATS = 0x02;
    public static final int MSG_CMD_CHANNEL_BUMP = 0x10;
    public static final int MSG_CMD_CHANNEL_DISCONNECT = 0x11;
    public static final int MSG_CMD_CHANNEL_KEEP_YES = 0x12;
    public static final int MSG_CMD_CHANNEL_KEEP_NO = 0x13;
    public static final int MSG_CMD_SERVENT_DISCONNECT = 0x20;

    private Messenger mServiceMessenger;
    private ServiceHandler mServiceHandler;

    private AppPreferences mPreferences;
    private int mRunningPort;
    private NotificationHelper mNotificationHelper;
    private ServiceReceiver mServiceReceiver;
    /**
     * arg2=startId
     */
    private static final Object MSG_OBJ_CALL_STOP_SELF = new Object();

    /**
     * PeerCastServiceController からの
     * .sendCommand() .sendChannelCommand()      を処理するHandler。
     */
    private final class ServiceHandler extends Handler {
        public ServiceHandler(Looper looper) {
            super(looper);
        }

        @Override
        public void handleMessage(Message msg) {
            Message reply = obtainMessage(msg.what);
            Boolean result = null;
            Bundle data = new Bundle();
            switch (msg.what) {
                case MSG_GET_APPLICATION_PROPERTIES:
                    data = nativeGetApplicationProperties();
                    break;

                case MSG_GET_CHANNELS:
                    data = nativeGetChannels();
                    break;

                case MSG_GET_STATS:
                    data = nativeGetStats();
                    break;

                case MSG_CMD_CHANNEL_BUMP:
                case MSG_CMD_CHANNEL_DISCONNECT:
                case MSG_CMD_CHANNEL_KEEP_YES:
                case MSG_CMD_CHANNEL_KEEP_NO:
                    result = nativeChannelCommand(msg.what, msg.arg1);
                    break;

                case MSG_CMD_SERVENT_DISCONNECT:
                    result = nativeDisconnectServent(msg.arg1);
                    break;

                default:
                    Log.e(TAG, "Illegal value: msg.what=" + msg.what);
                    return;
            }

            if (msg.obj == MSG_OBJ_CALL_STOP_SELF) {
                stopSelf(msg.arg2);
                return;
            }

            if (msg.replyTo == null) {
                return;
            }

            if (result != null)
                data.putBoolean("result", result);

            reply.setData(data);
            try {
                msg.replyTo.send(reply);
            } catch (RemoteException e) {
                Log.e(TAG, "msg.replyTo.send(reply)", e);
            }
        }
    }

    @Override
    public void onCreate() {
        HandlerThread thread = new HandlerThread(TAG);
        thread.start();

        mServiceHandler = new ServiceHandler(thread.getLooper());
        mServiceMessenger = new Messenger(mServiceHandler);
        mServiceReceiver = new ServiceReceiver(this);
        mPreferences = AppPreferences.from(this);
    }

    /**
     * 通知バーのボタンのイベントを処理する。
     */
    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        // Log.d(TAG, "onStartCommand: " + intent);
        String act = intent.getAction();
        Message msg = mServiceHandler.obtainMessage(-1);

        if (act != null) {
            // 通知バーのボタンが押された時
            msg.arg1 = intent.getIntExtra("channel_id", -1);
            if (ACTION_BUMP.equals(act)) {
                msg.what = MSG_CMD_CHANNEL_BUMP;
            } else if (ACTION_DISCONNECT.equals(act)) {
                msg.what = MSG_CMD_CHANNEL_DISCONNECT;
            }
        }
        msg.obj = MSG_OBJ_CALL_STOP_SELF;
        msg.arg2 = startId;
        mServiceHandler.sendMessage(msg);
        return START_NOT_STICKY;
    }

    @Override
    public IBinder onBind(Intent i) {
        File filesDir = getFilesDir();

        synchronized (this) {
            if (mRunningPort == 0) {
                try {
                    HtmlResource res = new HtmlResource(this);
                    if (!res.isInstalled())
                        res.doExtract();
                } catch (IOException e) {
                    Log.e(TAG, "html-dir install failed.", e);
                    return null;
                }
                mRunningPort = nativeStart(
                        new File(filesDir, "peercast.ini").getAbsolutePath(),
                        filesDir.getAbsolutePath()
                );
                mNotificationHelper = new NotificationHelper(this);

                if (mPreferences.isUPnPEnabled())
                    mServiceReceiver.registerConnectivityReceiver();

                if (mPreferences.isPowerSaveMode())
                    mServiceReceiver.registerScreenOffReceiver();
            }
        }

        return mServiceMessenger.getBinder();
    }


    @Override
    public boolean onUnbind(Intent intent) {
        return false;
    }

    @Override
    public void onDestroy() {
        stopForeground(true);

        mServiceReceiver.unregisterAll();

        mServiceHandler.getLooper().quit();
        if (mNotificationHelper != null)
            mNotificationHelper.quit();

        if (mRunningPort > 0 &&
                mPreferences.isUPnPEnabled() &&
                mPreferences.isUPnPCloseOnExit()) {
            Intent intent = new Intent(this, PecaPortService.class);
            intent.putExtra("close", mRunningPort);
            startService(intent);
        }

        nativeQuit();
    }

    /**
     * ネイティブ側から呼ばれる。<br>
     * <p>
     * AndroidPeercastApp::notifyMessage( ServMgr::NOTIFY_TYPE tNotify, const
     * char *message)
     *
     * @see #NOTIFY_CHANNEL_START
     * @see #NOTIFY_CHANNEL_UPDATE
     * @see #NOTIFY_CHANNEL_STOP
     */
    private void notifyMessage(int notifyType, String message) {
        if (BuildConfig.DEBUG) {
            Log.i(TAG, "notifyMessage: " + notifyType + ", " + message);
        }
    }

    private static final int NOTIFY_CHANNEL_START = 0;
    private static final int NOTIFY_CHANNEL_UPDATE = 1;
    private static final int NOTIFY_CHANNEL_STOP = 2;

    /**
     * ネイティブ側から呼ばれる。<br>
     * <p>
     * <pre>
     * AndroidPeercastApp::
     *   channelStart(ChanInfo *info)
     *   channelUpdate(ChanInfo *info)
     *   channelStop(ChanInfo *info)
     * </pre>
     *
     * @see #NOTIFY_CHANNEL_START
     * @see #NOTIFY_CHANNEL_UPDATE
     * @see #NOTIFY_CHANNEL_STOP
     * *
     */
    private void notifyChannel(int notifyType, Bundle bInfo) {
        if (BuildConfig.DEBUG) {
            String s = bInfo.getString("id");
            for (String key : bInfo.keySet())
                s += ", " + key + "=" + bInfo.get(key);
            Log.i(TAG, "notifyType_" + notifyType + ": " + s);
        }
    }

    /**
     * PeerCastを開始します。
     *
     * @param iniPath     peercast.iniのパス(要・書き込み可能)
     * @param resourceDir htmlディレクトリのある場所
     * @return 動作ポート
     */
    private native int nativeStart(String iniPath, String resourceDir);

    /**
     * PeerCastを終了します。
     */
    private native void nativeQuit();

    /**
     * 現在アクティブなChannel情報を取得します。
     *
     * @return Bundle。1つも無い場合はnull。2つ目以降はnextキーにリンクリスト形式で収納。
     * @see Channel#fromNativeResult(Bundle)
     */
    native Bundle nativeGetChannels();

    /**
     * 通信量の状態を取得します。
     *
     * @return Bundle
     * @see Stats#fromNativeResult(Bundle)
     */
    native Bundle nativeGetStats();

    /**
     * 動作中のポート番号などを取得します。
     *
     * @return Bundle
     */
    native Bundle nativeGetApplicationProperties();

    /**
     * チャンネルの再接続、キープの有無を操作する。
     *
     * @return 成功した場合 true
     */
    native boolean nativeChannelCommand(int cmdType,
                                        int channel_id);

    /**
     * 指定したサーヴァントを切断する。
     *
     * @return 成功した場合 true
     */
    native boolean nativeDisconnectServent(int servent_id);

    /**
     * クラス初期化に呼ぶ。
     */
    private static native void nativeClassInit();

    static {
        System.loadLibrary("peercast");
        nativeClassInit();
    }

    int getRunningPort() {
        return mRunningPort;
    }

    private static class HtmlResource {
        private final File mDataDir;
        //解凍済みを示す.IM45フォルダ
        private final File mInstalled;
        private final AssetManager mAssetManager;

        HtmlResource(Context c) {
            mDataDir = c.getFilesDir();
            mInstalled = new File(mDataDir, ".IM45");
            mAssetManager = c.getAssets();
        }

        /**
         * peca.zipからhtmlフォルダを解凍して /data/data/org.peercast.core/ にインストール。<br>
         * 成功した場合は解凍済みを示す .IM45フォルダを作成する。
         *
         * @throws IOException
         */
        public void doExtract() throws IOException {
            ZipInputStream zipIs = new ZipInputStream(mAssetManager.open("peca.zip"));
            try {
                ZipEntry entry;
                while ((entry = zipIs.getNextEntry()) != null) {
                    File file = new File(mDataDir, entry.getName());
                    if (entry.isDirectory())
                        continue;

                    Log.i(TAG, "Extract resource -> " + file.getAbsolutePath());
                    OutputStream os = null;
                    try {
                        os = FileUtils.openOutputStream(file);
                        IOUtils.copy(zipIs, os);
                    } finally {
                        IOUtils.closeQuietly(os);
                        zipIs.closeEntry();
                    }
                }
            } finally {
                IOUtils.closeQuietly(zipIs);
            }

            mInstalled.mkdir();
        }

        public boolean isInstalled() {
            return mInstalled.exists();
        }
    }
}

File: app/src/main/java/org/peercast/core/PeerCastService.kt
package org.peercast.core

/**
 * (c) 2014, T Yoshizawa
 * Dual licensed under the MIT or GPL licenses.
 */

import android.app.Service
import android.content.Intent
import android.os.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import org.koin.android.ext.android.inject
import org.peercast.core.lib.JsonRpcConnection
import org.peercast.core.lib.PeerCastController
import org.peercast.core.lib.PeerCastRpcClient
import org.peercast.core.lib.internal.PeerCastNotification
import org.peercast.core.lib.notify.NotifyChannelType
import org.peercast.core.util.AssetUnzip
import org.peercast.core.util.NotificationHelper
import org.peercast.pecaport.PecaPort
import timber.log.Timber
import java.io.File
import java.io.IOException
import kotlin.coroutines.CoroutineContext

class PeerCastService : Service(), CoroutineScope, Handler.Callback {

    private val serviceHandler = Handler(Looper.getMainLooper(), this)
    private lateinit var serviceMessenger: Messenger

    private val appPrefs by inject<AppPreferences>()
    private lateinit var notificationHelper: NotificationHelper
    private val job = Job()

    override val coroutineContext: CoroutineContext
        get() = job + Dispatchers.IO

    override fun onCreate() {
        serviceMessenger = Messenger(serviceHandler)
        notificationHelper = NotificationHelper(this, appPrefs)

        //解凍済みを示す空フォルダ ex: 3.0.0-YT28
        val extracted = File(filesDir, "${BuildConfig.VERSION_NAME}-${BuildConfig.YT_VERSION}")
        if (!extracted.exists()) {
            try {
                AssetUnzip.doExtract(this@PeerCastService, "peca-yt.zip", filesDir)
                extracted.mkdir()
            } catch (e: IOException) {
                Timber.e(e, "html-dir install failed.")
            }
        }

        nativeStart(filesDir.absolutePath, appPrefs.port)

        if (appPrefs.isUPnPEnabled) {
            PecaPort.openPort(this, appPrefs.port)
        }
    }

    override fun handleMessage(msg: Message): Boolean {
        val reply = serviceHandler.obtainMessage(msg.what)
        when (msg.what) {
            PeerCastController.MSG_GET_APPLICATION_PROPERTIES -> {
                reply.data.putInt("port", appPrefs.port)
            }

            //v3.1で廃止
            PeerCastController.MSG_PEERCAST_STATION_RPC -> {
                reply.data.putString(
                        PeerCastController.EX_RESPONSE,
                        """{"error":{"message":"obsoleted api"},"jsonrpc":"2.0"}"""
                )
            }
            //v3.0で廃止
            PeerCastController.MSG_GET_CHANNELS,
            PeerCastController.MSG_GET_STATS,
            PeerCastController.MSG_CMD_CHANNEL_BUMP,
            PeerCastController.MSG_CMD_CHANNEL_DISCONNECT,
            PeerCastController.MSG_CMD_CHANNEL_KEEP_YES,
            PeerCastController.MSG_CMD_CHANNEL_KEEP_NO,
            PeerCastController.MSG_CMD_SERVENT_DISCONNECT -> {
                Timber.e("Obsoleted API: msg.what=${msg.what}")
                return false
            }
            else -> {
                Timber.e("Illegal value: msg.what=${msg.what}")
                return false
            }
        }
        try {
            msg.replyTo?.send(reply)
        } catch (e: RemoteException) {
            Timber.e(e, "msg.replyTo.send(reply)")
        }
        return true
    }

    /**
     * 通知バーのボタンのイベントを処理する。
     */
    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        Timber.d("onStartCommand(intent=$intent, flags=$flags, startId=$startId)")

        val channelId = intent?.getStringExtra(EX_CHANNEL_ID)
                ?: return super.onStartCommand(intent, flags, startId)
        val conn = JsonRpcConnection(port = appPrefs.port)
        val client = PeerCastRpcClient(conn)
        val f = when (intent.action) {
            ACTION_BUMP_CHANNEL -> client::bumpChannel
            ACTION_STOP_CHANNEL -> client::stopChannel
            else -> return super.onStartCommand(intent, flags, startId)
        }
        launch {
            runCatching {
                f(channelId)
            }.onFailure(Timber::e)
        }
        return START_NOT_STICKY
    }

    override fun onBind(intent: Intent): IBinder? {
        return serviceMessenger.binder
    }

    override fun onUnbind(intent: Intent): Boolean {
        return false
    }

    override fun onDestroy() {
        job.cancel()

        if (appPrefs.isUPnPEnabled &&
                appPrefs.isUPnPCloseOnExit) {
            PecaPort.closePort(this, appPrefs.port)
        }

        stopForeground(true)
        nativeQuit()
    }

    /**
     * ネイティブ側から呼ばれる。
     * @see org.peercast.core.lib.notify.NotifyMessageType
     */
    @Suppress("unused")
    private fun notifyMessage(notifyType: Int, message: String) {
//        if (BuildConfig.DEBUG) {
//            Timber.i("notifyMessage: $notifyType, $message")
//        }
        PeerCastNotification.sendBroadCastNotifyMessage(this, notifyType, message)
    }


    /**
     * ネイティブ側から呼ばれる。
     * @see org.peercast.core.lib.notify.NotifyChannelType
     */
    @Suppress("unused")
    private fun notifyChannel(notifyType: Int, chId: String, jsonChannelInfo: String) {
        val chInfo = PeerCastNotification.jsonToChannelInfo(jsonChannelInfo) ?: return
        when (notifyType) {
            NotifyChannelType.Start.nativeValue ->
                notificationHelper.start(chId, chInfo)
            NotifyChannelType.Update.nativeValue ->
                notificationHelper.update(chId, chInfo)
            NotifyChannelType.Stop.nativeValue ->
                notificationHelper.remove(chId)
            else -> throw IllegalArgumentException()
        }
        PeerCastNotification.sendBroadCastNotifyChannel(this, notifyType, chId, jsonChannelInfo)
        //Timber.d("$notifyType $chId $chInfo ${Thread.currentThread()}")
    }

    /**
     * PeerCastを開始します。
     *
     * @param filesDirPath     Context.getFilesDir()
     * @param port 動作ポート
     */
    private external fun nativeStart(filesDirPath: String, port: Int)

    /**
     * PeerCastを終了します。
     */
    private external fun nativeQuit()

    companion object {
        // 通知ボタンAction
        const val ACTION_BUMP_CHANNEL = "org.peercast.core.ACTION.bumpChannel"
        const val ACTION_STOP_CHANNEL = "org.peercast.core.ACTION.stopChannel"

        /**(String)*/
        const val EX_CHANNEL_ID = "channelId"

        /**
         * クラス初期化に呼ぶ。
         */
        @JvmStatic
        private external fun nativeClassInit()

        init {
            System.loadLibrary("peercast")
            nativeClassInit()
        }
    }

}


